### PR TITLE
Feat/8599 update saved addresses

### PR DESCRIPTION
**Pull request by IvanBelyakoff**
Update saved addresses DB to store ENS, isTest flag and chains
prefixes. Changed primary keys and API methods.

Fixes #8599

Important changes:
- [x] Removed ChainID from saved addresses DB as not needed and not really used
- [x] Decision made to not support same addresses in DB with different names for multiple reasons:

1. Using name + is_test fields as PRIMARY KEY will not work in case we want to change the name of the saved address - a new records is created.
2. Adding a new GUID column as PRIMARY KEY makes DB migration more complex since it must be NOT NULL and old database does not have that column, so need extra efforts to fill that column and pass GUID from UI for new records.
3. Not really an important feature

- [x] Replaced a tmp DB file with in memory DB to speed up testing
- [x] Fixed tests and added new ones 

Closes #8599

### CODE REVIEW

**Comment by ghost** [comment]
### Pull Request Checklist
- [x] Have you updated the documentation, if impacted (e.g. [docs.status.im](https://status.im/docs/))?
- [ ] Have you tested changes with mobile?
- [x] Have you tested changes with desktop?

File: services/wallet/saved_addresses.go [hunk 1]
```diff
@@ -15,16 +15,18 @@ type savedAddressMeta struct {
...
 type SavedAddress struct {
 	Address common.Address `json:"address"`
-	// TODO: Add Emoji and Networks
-	// Emoji    string         `json:"emoji"`
+	// TODO: Add Emoji
+	// Emoji	string		   `json:"emoji"`
 	Name      string `json:"name"`
 	Favourite bool   `json:"favourite"`
-	ChainID   uint64 `json:"chainId"`
+	Networks  string `json:"chains"` // used with address only, not with ENSName
```
Comment (alaibe) [inline services/wallet/saved_addresses.go]: nit, why naming this Networks, instead of ChainIDs, this is what we use in wallet?

File: services/ext/api.go [hunk 2]
```diff
@@ -914,19 +914,11 @@ func (api *PublicAPI) DestroyWalletConnectSession(ctx context.Context, PeerID st
...
 	return api.service.messenger.UpsertSavedAddress(ctx, sa)
 }
 
-func (api *PublicAPI) DeleteSavedAddress(ctx context.Context, chainID uint64, address ethcommon.Address) error {
-	if chainID == 0 {
-		chainID = api.service.rpcClient.UpstreamChainID
-	}
-
-	return api.service.messenger.DeleteSavedAddress(ctx, chainID, address)
+func (api *PublicAPI) DeleteSavedAddress(ctx context.Context, address ethcommon.Address, ens string, isTest bool) error {
```
Comment (alaibe) [inline services/ext/api.go:924]: isTest can be inferred from the settings, so that u don't need to expose this in the API?

File: services/wallet/api.go [hunk 3]
```diff
@@ -214,24 +214,21 @@ func (api *API) DeleteCustomTokenByChainID(ctx context.Context, chainID uint64,
...
-	if sa.ChainID == 0 {
-		sa.ChainID = api.s.rpcClient.UpstreamChainID
-	}
 	_, err := api.s.savedAddressesManager.UpdateMetadataAndUpsertSavedAddress(sa)
 	log.Debug("result from database for create or edit saved address", "err", err)
 	return err
 }
 
-func (api *API) DeleteSavedAddress(ctx context.Context, address common.Address) error {
+func (api *API) DeleteSavedAddress(ctx context.Context, address common.Address, ens string, isTest bool) error {
```
Comment (alaibe) [inline services/wallet/api.go:217]: isTest can be inferred right?

File: appdatabase/migrations/sql/1676279805_add_chains_ens_istest_to_saved_addresses.up.sql [hunk 4]
```diff
@@ -0,0 +1,17 @@
+CREATE TABLE IF NOT EXISTS saved_addresses_new (
+  address VARCHAR NOT NULL,
+  name TEXT NOT NULL,
+  favourite BOOLEAN NOT NULL DEFAULT FALSE,
+  removed BOOLEAN NOT NULL DEFAULT FALSE,
+  update_clock INT NOT NULL DEFAULT FALSE,
+  chains VARCHAR DEFAULT "",
```
Comment (alaibe) [inline appdatabase/migrations/sql/1676279805_add_chains_ens_istest_to_saved_addresses.up.sql]: nit: rename to chainIDs ?

File: services/wallet/saved_addresses.go (hunk 1 above)
Comment (IvanBelyakoff) [reply services/wallet/saved_addresses.go]: Initially I did not remove ChainID, so I thought maybe it is not good to have ChainID and ChainIDs, ans since in the code - for example in database ChainId was called network_id, hence the name.
But the question is the same as below - the field stores chains short names as entered by user: "eth:opt:".
What do you think, ChainShortNames is good enough? If call it ChainIDs, I would expect 1 for Eth, 42161 for Arbitrum

File: services/ext/api.go (hunk 2 above)
Comment (IvanBelyakoff) [reply services/ext/api.go:924]: I was thinking about it, but I thought maybe it is more reasonable to not add an extra dependency on other service at this low level unless necessary to avoid tighter coupling and have it as a param from higher level instead. So that was more like an architectural decision.
Do you think it is better to have retrieve the value from `multiaccounts` service?

File: appdatabase/migrations/sql/1676279805_add_chains_ens_istest_to_saved_addresses.up.sql (hunk 4 above)
Comment (IvanBelyakoff) [reply appdatabase/migrations/sql/1676279805_add_chains_ens_istest_to_saved_addresses.up.sql]: what about `chain_ids` - not to mix camel and snake cases in the table. Is it good?

Comment (IvanBelyakoff) [reply appdatabase/migrations/sql/1676279805_add_chains_ens_istest_to_saved_addresses.up.sql]: Btw, they are prefixes as this is a string. So the values are for example. "eth:opt:"
that is why I did not use "ids" word. Or you think it is still good?

File: services/wallet/saved_addresses.go (hunk 1 above)
Comment (alaibe) [reply services/wallet/saved_addresses.go]: yes for: ChainShortNames

File: appdatabase/migrations/sql/1676279805_add_chains_ens_istest_to_saved_addresses.up.sql (hunk 4 above)
Comment (alaibe) [reply appdatabase/migrations/sql/1676279805_add_chains_ens_istest_to_saved_addresses.up.sql]: as u suggested before: chain_short_names makes sense

File: services/wallet/api.go (hunk 3 above)
Comment (alaibe) [reply services/wallet/api.go:217]: maybe not in the API but savedAddressesManager should have access to the multiaccounts

File: services/ext/api.go (hunk 2 above)
Comment (IvanBelyakoff) [reply services/ext/api.go:924]: @alaibe there is one case I discovered when adding more sync tests, why I decided to drop the fix to get isTest flag from multiaccounts service. Here it is:
- First device has test mode enabled
- Second device has normal mode enabled
- when syncing saved addresses happens, isTest flag is taken from accounts service on both devices and they are different, so syncing fails in case of deletion of saved address. When it is passed along with other DB primary keys, it works as expected.
Please let me know if you agree to keep isTest param in parameters list

**Review by stefandunca — APPROVED** [review]
Great work!

**Review by alaibe — APPROVED** [review]

File: appdatabase/migrations/sql/1676279805_add_chains_ens_istest_to_saved_addresses.up.sql [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+  update_clock INT NOT NULL DEFAULT FALSE,
+  chain_short_names VARCHAR DEFAULT "",
+  ens_name VARCHAR DEFAULT "",
+  is_test BOOLEAN DEFAULT FALSE,
+  PRIMARY KEY (address, ens_name, is_test)
+) WITHOUT ROWID;
+
+INSERT INTO saved_addresses_new (address, name, favourite, removed, update_clock)
+   SELECT address, name, favourite, removed, update_clock FROM saved_addresses;
+ALTER TABLE saved_addresses RENAME TO saved_addresses_old;
```
Comment (saledjenic) [inline appdatabase/migrations/sql/1676279805_add_chains_ens_istest_to_saved_addresses.up.sql]: @IvanBelyakoff you don't need this line, you could do `DROP TABLE saved_addresses;` here and remove line 17.

File: protocol/protobuf/pairing.proto [hunk 6]
```diff
@@ -223,9 +223,11 @@ message SyncSavedAddress {
   bytes  address = 1;
   string name = 2;
   bool   favourite = 3;
-  uint64 chain_id = 4;
```
Comment (saledjenic) [inline protocol/protobuf/pairing.proto:231]: @IvanBelyakoff in such cases it's better to use `uint64 chain_id = 4 [deprecated = true];` or `reserved 4;` then deleting it, cause that will prevent users from using it.

File: protocol/protobuf/pairing.proto [hunk 7]
```diff
@@ -223,9 +223,11 @@ message SyncSavedAddress {
...
   bool   removed = 5;
   uint64 update_clock = 7;
```
Comment (saledjenic) [inline protocol/protobuf/pairing.proto:233]: @IvanBelyakoff maybe we could use just `clock`, just for the sake of consistency, to stick to the way how it's done for other messages.

File: services/wallet/saved_addresses.go [hunk 8]
```diff
@@ -54,8 +56,13 @@ func getSavedAddressesFromDBRows(rows *sql.Rows) ([]SavedAddress, error) {
 	return addresses, nil
 }
 
-func (sam *SavedAddressesManager) GetSavedAddressesForChainID(chainID uint64) ([]SavedAddress, error) {
-	rows, err := sam.db.Query(fmt.Sprintf("SELECT %s FROM saved_addresses WHERE network_id = ? AND removed != 1", rawQueryColumnsOrder), chainID)
+func (sam *SavedAddressesManager) getSavedAddresses(condition string) ([]SavedAddress, error) {
+	var whereCondition string
+	if condition != "" {
+		whereCondition = "WHERE " + condition
```
Comment (saledjenic) [inline services/wallet/saved_addresses.go]:
```suggestion
whereCondition = fmt.Sprintf("WHERE %s", condition)
```

File: services/wallet/saved_addresses_test.go [hunk 9]
```diff
@@ -14,49 +13,50 @@ import (
...
+	ensMember int = iota
+	isTestMember
+	addressMember
+)
+
 func setupTestSavedAddressesDB(t *testing.T) (*SavedAddressesManager, func()) {
-	tmpfile, err := ioutil.TempFile("", "wallet-saved_addresses-tests-")
-	require.NoError(t, err)
-	db, err := appdatabase.InitializeDB(tmpfile.Name(), "wallet-saved_addresses-tests", sqlite.ReducedKDFIterationsNumber)
+	db, err := appdatabase.InitializeDB(sqlite.InMemoryPath, "wallet-saved_addresses-tests", sqlite.ReducedKDFIterationsNumber)
```
Comment (saledjenic) [inline services/wallet/saved_addresses_test.go:23]: @IvanBelyakoff why did you change this?

Not sure, but maybe this could work as well:
```
func setupTestSavedAddressesDB(t *testing.T) (* SavedAddressesManager, func()) {
	db, stop, err := appdatabase.SetupTestSQLDB("wallet-saved_addresses-tests")
	if err != nil {
		require.NoError(t, stop())
	}
	require.NoError(t, err)

        return &SavedAddressesManager{db}, func() {
		require.NoError(t, stop())
	}
}
```

File: appdatabase/migrations/sql/1676279805_add_chains_ens_istest_to_saved_addresses.up.sql (hunk 5 above)
Comment (IvanBelyakoff) [reply appdatabase/migrations/sql/1676279805_add_chains_ens_istest_to_saved_addresses.up.sql]: Thank you for the feedback. That is just normal precautions when you want to overwrite something - make a backup, overwrite, and if that goes well - remove the backup.
In this case we might want to skip the backup part. I can do that

File: services/wallet/saved_addresses_test.go (hunk 9 above)
Comment (IvanBelyakoff) [reply services/wallet/saved_addresses_test.go:23]: I did that because file creation and working with DB on hard drive takes a lot of time and I have found that there is already an option to use in-memory DB. So I think for test cases we don't need a DB on hard drive. And my tests show 10 times performance gain (5 sec for wallet tests -> 0.5).

Comment (IvanBelyakoff) [reply services/wallet/saved_addresses_test.go:23]: SetupTestSQLDB - creates a file as well, which is slow. Do we really need to create a DB on hard drive for test suites?

File: protocol/protobuf/pairing.proto (hunk 7 above)
Comment (IvanBelyakoff) [reply protocol/protobuf/pairing.proto:233]: you mean `clock` instead of `update_clock` ? I think I will have to add something like `create_clock` soon, to be able to sort records by creation date (requirement from Ben was that a new saved address must popup at the top of the list, and update_clock will not work for this purpose, as it will change the order of the items in the list for every single change on any record -like clicking `favourite` button). So I think just single "clock" will not be enough anyways

File: appdatabase/migrations/sql/1676279805_add_chains_ens_istest_to_saved_addresses.up.sql (hunk 5 above)
Comment (saledjenic) [reply appdatabase/migrations/sql/1676279805_add_chains_ens_istest_to_saved_addresses.up.sql]: In general you don't have to do anything, but just to clarify:

- what's done:
from `saved_addresses` to `saved_addresses_new`
rename `saved_addresses` to `saved_addresses_old`
rename `saved_addresses_new` to  `saved_addresses`
delete `saved_addresses_old`

- what I suggested:
from `saved_addresses` to `saved_addresses_new`
delete `saved_addresses`
rename `saved_addresses_new` to  `saved_addresses`

File: protocol/protobuf/pairing.proto (hunk 7 above)
Comment (saledjenic) [reply protocol/protobuf/pairing.proto:233]: If that's the request than ok. Will we have to select among sort criteria, like `newest first`/`most recently updated first`? If there are both options, then ok, but if there is no options I guess it should be sort by `most recently updated first`.

**Review by saledjenic — APPROVED** [review]

File: protocol/protobuf/pairing.proto (hunk 7 above)
Comment (stefandunca) [reply protocol/protobuf/pairing.proto:233]: If I remember correctly, for other implementations I had as an example, the clock is coming from the messenger (lamport clock maybe?). In the case of the saved address, this is the user's wall clock from when the entry was updated independent of sync or messenger. It is used to solve editing conflicts.

File: services/wallet/saved_addresses_test.go (hunk 9 above)
Comment (stefandunca) [reply services/wallet/saved_addresses_test.go:23]: Great thinking @IvanBelyakoff. Usually, the file I/O is much slower than the random memory access, which should provide considerable gains in the development environment and a good speed up in the CI. Most of the tests, should not require file-based storage (for sure, the ones that don't deal with user folder management).

Let's bring this matter to the `status-go` channel and find out if anybody notices any issue with bringing this improvement to all tests (`SetupTestSQLDB ` and general setup entries)
